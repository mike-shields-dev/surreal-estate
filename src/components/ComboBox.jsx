import React from "react";
import PropTypes from "prop-types";
import styles from "../styles/ComboBox.module.css";

const Combobox = ({ field, onChange, value, values }) => {
  return (
    <>
      <label htmlFor={field} className={styles.combobox__label}>
        {field}
      </label>
      <select id={field} {...{ name: field, value, onChange }}>
        {values.map((val) => {
          const id = `${val}-${field}-option`;
          return (
            <option key={id} value={val}>
              {val}
            </option>
          );
        })}
      </select>
    </>
  );
};

Combobox.propTypes = {
  field: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  values: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default Combobox;
