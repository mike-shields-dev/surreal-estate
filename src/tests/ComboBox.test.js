import React from "react";
import { fireEvent, render, screen, within } from "@testing-library/react";
import ComboBox from "../components/ComboBox";

const values = ["Manchester", "Leeds", "Sheffield", "Liverpool"];
const props = {
  field: "city",
  onChange: jest.fn(),
  value: values[0],
  values,
};

const renderComboBox = () => render(<ComboBox {...props} />);

describe("ComboBox", () => {
  it("matches snapshot", () => {
    const { asFragment } = renderComboBox();

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders select input with associated label and given props", () => {
    renderComboBox();

    const selectInput = screen.getByLabelText(props.field);

    expect(selectInput.getAttribute("id")).toEqual(props.field);
    expect(selectInput.getAttribute("name")).toEqual(props.field);
    expect(selectInput).toHaveValue(props.value);
  });

  it("renders the given values as options", () => {
    renderComboBox();

    values.forEach((value) => {
      expect(
        within(screen.getByRole("combobox", { name: props.field })).getByText(
          value
        )
      ).toBeInTheDocument();
    });
  });

  it("invokes the onChange callback", async () => {
    renderComboBox();

    fireEvent.change(screen.getByRole("combobox", { name: props.field }), {
      target: { value: values[3] },
    });

    expect(props.onChange).toHaveBeenCalled();
  });
});
