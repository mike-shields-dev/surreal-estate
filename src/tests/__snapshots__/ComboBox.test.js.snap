// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComboBox matches snapshot 1`] = `
<DocumentFragment>
  <label
    class="combobox__label"
    for="city"
  >
    city
  </label>
  <select
    id="city"
    name="city"
  >
    <option
      value="Manchester"
    >
      Manchester
    </option>
    <option
      value="Leeds"
    >
      Leeds
    </option>
    <option
      value="Sheffield"
    >
      Sheffield
    </option>
    <option
      value="Liverpool"
    >
      Liverpool
    </option>
  </select>
</DocumentFragment>
`;
